### PR TITLE
[ARCTIC-1698] Fix delete change ttl files will commit empty snapshot

### LINKED ISSUE #1698: [Bug]: there are some empty commit
### What happened?

there are some empty commit when no data input
image

### Affects Versions

master

### What engines are you seeing the problem on?

Optimizer

### How to reproduce

table:   mixed hive table, with primary key , no partition key
data:  insert data from a flink datagen source.
optimizer:  Flink optimizer

After stopping the upstream source for more than a day, it was discovered that there were some empty commits.

**I'm not sure if I can reproduce it reliably.**

image

image

### Relevant log output

_No response_

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by HuangFru**
## Why are the changes needed?
Fix #1698.
Fix related problems mentioned in #1656.

## Brief change log

  - *Use entry scan to plan and filter files that are expired.*
  - *Use entries' sequence numbers to determine whether they should be deleted.*
  - *Make filter condition in 'tryEvaluatingPendingInput' instead of 'execute'.*
  - *Combine the snapshot-keep-time with ttl in the change store.*
  - *Complete UT test.*

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by HuangFru** [comment]
cc @wangtaohz.
By the way, I found some grammatical errors in the pull request template and fixed them.

**Comment by HuangFru** [comment]
Fix problems mentioned in #1656 by the way.

File: ams/server/src/main/java/com/netease/arctic/server/table/executor/TableRuntimeRefreshExecutor.java [hunk 1]
```diff
@@ -52,12 +52,16 @@ public void handleStatusChanged(TableRuntime tableRuntime, OptimizingStatus orig
...
-    if (evaluator.isNecessary()) {
-      OptimizingEvaluator.PendingInput pendingInput = evaluator.getPendingInput();
-      logger.debug("{} optimizing is necessary and get pending input {}", tableRuntime.getTableIdentifier(),
-          pendingInput);
-      tableRuntime.setPendingInput(pendingInput);
+    if (tableRuntime.isOptimizingEnabled() && !tableRuntime.getOptimizingStatus().isProcessing()) {
+      OptimizingEvaluator evaluator = new OptimizingEvaluator(tableRuntime, table);
+      if (evaluator.isNecessary()) {
+        OptimizingEvaluator.PendingInput pendingInput = evaluator.getPendingInput();
+        if (logger.isDebugEnabled()) {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/table/executor/TableRuntimeRefreshExecutor.java:59]: We no longer judge whether the debug mode is enabled.
reference: https://github.com/NetEase/arctic/issues/1570

File: ams/server/src/main/java/com/netease/arctic/server/table/executor/SnapshotsExpiringExecutor.java [hunk 2]
```diff
@@ -258,31 +255,26 @@ public static void expireSnapshots(
...
-  public static List<DataFile> getClosestExpireDataFiles(UnkeyedTable changeTable, Snapshot closestExpireSnapshot) {
-    List<DataFile> changeTTLDataFiles = new ArrayList<>();
-    long recentExpireSnapshotId = closestExpireSnapshot.snapshotId();
-    try (CloseableIterable<FileScanTask> fileScanTasks = changeTable.newScan()
-        .useSnapshot(recentExpireSnapshotId)
-        .planFiles()) {
-      fileScanTasks.forEach(fileScanTask -> changeTTLDataFiles.add(fileScanTask.file()));
+    try (CloseableIterable<IcebergFileEntry> entries = entriesScan.entries()) {
+      entries.forEach(entry -> {
+        if (changeTable.snapshot(entry.getSnapshotId()).timestampMillis() <= ttlPoint) {
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/table/executor/SnapshotsExpiringExecutor.java:266]: It brings NPE if the snapshot is expired. We should also collect the file to `changeTTLFileEntries` if its snapshot does not exist.

**Review by wangtaohz — CHANGES_REQUESTED** [review]
I left some comments. Please take a look. @HuangFru

**Comment by HuangFru** [comment]
> I left some comments. Please take a look. @HuangFru

Fixed in the new commit.

**Review by wangtaohz — APPROVED** [review]
LGTM

**Comment by HuangFru** [comment]
`snapshot.change.keep.minutes` is merged into `change.data.ttl.minutes` in the new commit.

File: ams/server/src/main/java/com/netease/arctic/server/table/executor/SnapshotsExpiringExecutor.java [hunk 3]
```diff
@@ -258,31 +255,34 @@ public static void expireSnapshots(
...
-  public static Snapshot getClosestExpireSnapshot(UnkeyedTable changeTable, long ttl) {
+  public static Snapshot getClosestExpireSnapshot(UnkeyedTable changeTable, long ttlPoint) {
     if (changeTable.snapshots() == null) {
       return null;
     }
     return Streams.stream(changeTable.snapshots())
-        .filter(snapshot -> snapshot.timestampMillis() <= ttl)
-        .min(Comparator.comparingLong(snapshot -> Math.abs(ttl - snapshot.timestampMillis())))
+        .filter(snapshot -> snapshot.timestampMillis() < ttlPoint)
+        .min(Comparator.comparingLong(snapshot -> Math.abs(ttlPoint - snapshot.timestampMillis())))
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/table/executor/SnapshotsExpiringExecutor.java:264]: We can simplify the code and just use the `snapshot.timestampMillis()` to compare.

Comment (HuangFru) [reply ams/server/src/main/java/com/netease/arctic/server/table/executor/SnapshotsExpiringExecutor.java:264]: I get it.

**Review by wangtaohz — APPROVED** [review]
